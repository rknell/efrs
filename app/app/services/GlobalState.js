import { observable, observe, spy } from 'mobx'
import { Alert, AsyncStorage, Linking } from 'react-native'
import PushNotifications from './PushNotifications'
import Translate from '../services/Translate'

class GlobalState {
  // apiPath = 'http://192.168.1.8:1676/api'
  apiPath = 'https://www.efrs.io/api'
  // apiPath = 'http://localhost:1676/api'

  @observable model = {
    userToken: null,
    user: null,
    userPermissions: {
      permissions: {},
      groups: []
    },
    canPop: false,
    forms: [],
    cachedCalls: {
      get: {},
      post: []
    },
    isBusy: 0,
    pushToken: null
  }

  current = () => {
    return {
      user: this.model.user ? this.model.user._id : 'no user',
      group: this.model.userPermissions.groups.join(',')
    }
  }

  navigation = {
    resetTo: () => {
      console.log('RESETTING NAVIGATION')
    },
    push: () => {
      console.log('PUSHING NAVIGATION')
    }
  }

  saveSetting (key, value) {
    return AsyncStorage.setItem(`@EFRS.AppSettings:${key}`, value)
  }

  getSetting (key) {
    return AsyncStorage.getItem(`@EFRS.AppSettings:${key}`)
  }

  constructor () {
    spy(event => {
      if (event.type === 'action' || event.type === 'update') {
        // console.log(event);
      }
    })

    this.getSetting('formsData').then(data => {
      if (data) this.model.forms = JSON.parse(data)
    })

    const PNObj = new PushNotifications(this._setPushToken)
    this.pushNotifications = PNObj.messaging

    observe(this.model, 'user', () => {
      try {
        if (this.model.user && this.bugsnag) {
          this.bugsnag.setUser(this.model.user._id, `${this.model.user.firstName} ${this.model.user.lastName}`, this.model.user.email)
        }
      } catch (e) {
        console.error('could not set user for bugsnag', e)
      }
    })
  }

  _setPushToken = (token) => {
    this.model.pushToken = token
  }

  openUrl (url) {
    Linking.canOpenURL(url).then(supported => {
      if (!supported) {
        Alert.alert(Translate.do('Error'), Translate.do('Can\'t open this url on this device. Try installing an app that can handle this link type and try again.'))
      } else {
        return Linking.openURL(url)
      }
    }).catch(err => console.error('An error occurred', err))
  }
}

export default new GlobalState()
