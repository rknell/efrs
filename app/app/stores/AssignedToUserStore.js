import OfflineDataStore from './OfflineDataStore'
import { ResponseStoreItem } from './ResponseStore'

export default class AssignedToUserStore extends OfflineDataStore {
  constructor (networkRequestHandler) {
    super('assignedToUser', ResponseStoreItem, networkRequestHandler)
  }

  refresh () {
    this.isRefreshing = true
    return this.networkRequestHandler.call({ endpoint: `/response/assignedToMe` })
      .then(data => {
        if (data) {
          this.repopulateData(data)
          this.saveOffline()
        }
      })
      .finally(() => {
        this.isRefreshing = false
      })
  }
}
