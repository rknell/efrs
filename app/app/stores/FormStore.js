import OfflineDataStore, { OfflineDataItem } from './OfflineDataStore'
import { observable } from 'mobx'
import Translate from '../services/Translate'

export default class FormStore extends OfflineDataStore {
  constructor (networkRequestHandler) {
    super('form', FormStoreItem, networkRequestHandler)
  }
}

export class FormStoreItem extends OfflineDataItem {
  @observable name = ''
  @observable company = ''
  @observable createdBy = null
  @observable dateStarted = new Date()
  @observable restrictedGroups = []
  @observable headerImageUrl = ''
  @observable emailResponses = ''
  @observable emailToSupervisor = false
  @observable lastFieldId = 0
  @observable pages = [
    {
      name: Translate.do('Page') + ' 1',
      groups: observable([
        {
          fields: observable([])
        }
      ])
    }
  ]

  @observable currentVersion = 1

  clone = () => {
    const data = JSON.parse(JSON.stringify(this))
    delete data.createdBy
    delete data.createdAt
    delete data.updatedAt
    delete data._id
    data.name = this.name + ' (copy)'
    return this.store().create(data)
  }
}
