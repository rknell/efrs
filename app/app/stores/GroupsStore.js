import OfflineDataStore, { OfflineDataItem } from './OfflineDataStore'
import { observable } from 'mobx'

export default class GroupStore extends OfflineDataStore {
  constructor (networkRequestHandler) {
    super('groups', GroupStoreItem, networkRequestHandler)
  }

  refresh () {
    this.isRefreshing = true
    return this.networkRequestHandler.call({ endpoint: `/user/listGroups` })
      .then(data => {
        if (data) {
          this.repopulateData(data)
          this.saveOffline()
        }
      })
      .finally(() => {
        this.isRefreshing = false
      })
  }
}

export class GroupStoreItem extends OfflineDataItem {
  @observable _id = null
  @observable guid = null
  @observable name = ''
  @observable users = []
  @observable permissions = {}
}
