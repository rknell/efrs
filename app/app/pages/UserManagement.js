import React, { Component } from 'react'
import { Alert, RefreshControl, SectionList, Text, TouchableOpacity, View } from 'react-native'
import NavBar from '../components/NavBar'
import GlobalState from '../services/GlobalState'
import { computed, observable } from 'mobx'
import { observer } from 'mobx-react/native'
import GroupManagement from './GroupManagement'
import GlobalStyle from '../services/GlobalStyle'
import UserDetailView from '../components/UserDetailView'
import stores from '../stores'
import SearchBar from '../components/SearchBar'
import Translate from '../services/Translate'

@observer
class UserItem extends Component {
  getRelationshipUser (userId) {
    const user = stores.users.data.find(user => user._id === userId)
    if (user) {
      return user.username
    } else {
      return 'User not found'
    }
  }

  render () {
    return (
      <TouchableOpacity onPress={this.props.onPress}>
        <View style={{
          borderBottomWidth: 1,
          borderBottomColor: this.props.isLast ? 'transparent' : '#e7eaec',
          padding: 5,
          paddingLeft: 8
        }}>
          {(!!this.props.user.lastName || !!this.props.user.firstName) && <View style={{ flexDirection: 'row' }}>
            <Text style={{ fontWeight: 'bold' }}>
              {!!this.props.user.lastName && <Text>{this.props.user.lastName}, </Text>}
              {!!this.props.user.firstName && <Text>{this.props.user.firstName}</Text>}
            </Text>
          </View>}
          <Text>{this.props.user.username}</Text>
          {!!this.props.user.relationships && this.props.user.relationships.map(relationship =>
            <Text>
              <Text style={{ fontWeight: 'bold' }}>{relationship.name}</Text>
              &nbsp;- {this.getRelationshipUser(relationship.user)}
            </Text>)}
        </View>
      </TouchableOpacity>
    )
  }
}

@observer
class UserDetailPage extends Component {
  @observable state = {
    isUserValid: false
  }

  constructor (props) {
    super(props)

    if (this.props.canDelete) {
      this.pageActions.unshift({
        icon: 'trash',
        onPress: this.onDelete,
        label: 'Delete User'
      })
    }

    this.originalUser = JSON.stringify(this.props.user)
  }

  onSaveButton = () => {
    console.log('on save called')
    if (!this.state.isUserValid) {
      Alert.alert(Translate.do('Invalid user'), Translate.do('The user record is not valid, please correct the errors and try again.'))
    } else {
      GlobalState.navigation.pop()
      this.props.onSave(this.props.user)
    }
  }

  onDelete = () => {
    GlobalState.navigation.pop()
    this.props.onDelete(this.props.user)
  }

  onSaveBack = () => {
    this.props.onSave(this.props.user)
    GlobalState.navigation.pop()
  }

  onNoSave = () => {
    const originalUser = JSON.parse(this.originalUser)
    for (const key in originalUser) {
      this.props.user[key] = originalUser[key]
    }
    GlobalState.navigation.pop()
  }

  handleBack = () => {
    if (this.originalUser !== JSON.stringify(this.props.user)) {
      Alert.alert(Translate.do('Unsaved Changes'), Translate.do('The users has unsaved changes'), [
        {
          text: Translate.do('Don\'t Save'), onPress: this.onNoSave
        },
        {
          text: Translate.do('Save Changes'),
          onPress: this.onSaveBack
        }
      ])
    } else {
      GlobalState.navigation.pop()
    }
  }

  onValidationChanged = (value) => {
    this.state.isUserValid = value
  }

  pageActions = [
    {
      icon: 'save',
      onPress: this.onSaveButton,
      label: 'Save Changes'
    }
  ]

  render () {
    return <View style={{ flex: 1, backgroundColor: GlobalStyle.style.neutralColour }}>
      <NavBar title='User Details'
        isMenuDisabled
        pageActions={this.pageActions}
        handlePop={this.handleBack}
      />
      <UserDetailView user={this.props.user} onValidationChanged={this.onValidationChanged} />

    </View>
  }
}

@observer
class UserManagement extends Component {
  @observable state = {
    searchText: ''
  }

  constructor (props) {
    super(props)
    if (props.controller) {
      this.controller = props.controller
    } else {
      this.controller = new UserManagementController(props)
    }
  }

  _meetsFilter (string) {
    return !this.state.searchText || (!!string && string.toLowerCase().indexOf(this.state.searchText.toLowerCase()) > -1)
  }

  @computed get sections () {
    if (!stores.users.isRefreshing) {
      const sections = stores.users.data.reduce((result, item) => {
        let sectionCode
        if (item && item.lastName && item.lastName[0]) {
          sectionCode = item.lastName[0].toUpperCase()
        } else {
          sectionCode = ' NO NAME'
        }
        const section = result.find(section => section.title === sectionCode)
        if (this._meetsFilter(item.firstName) || this._meetsFilter(item.lastName) || this._meetsFilter(item.username) || this._meetsFilter(item.email)) {
          if (section) {
            section.data.push(item)
          } else {
            result.push({
              data: [item],
              title: sectionCode
            })
          }
        }

        return result
      }, [])

      sections.forEach(section => {
        section.data.sort((a, b) => {
          if (a.lastName < b.lastName) {
            return -1
          } else {
            return 1
          }
        })
      })

      sections.sort((a, b) => {
        if (a.title < b.title) {
          return -1
        } else {
          return 1
        }
      })
      return sections
    } else {
      return []
    }
  }

  searchBarChange = (value) => {
    this.state.searchText = value
  }

  render () {
    return (
      <View style={{ flex: 1, backgroundColor: GlobalStyle.style.neutralColour }}>
        <NavBar
          title='User Management'
          pageActions={[
            {
              icon: 'plus',
              onPress: this.controller.startCreateUser,
              label: 'Add user'
            },
            {
              icon: 'edit',
              onPress: this.controller.manageGroups,
              label: 'Manage Groups'
            }
          ]}
        />

        <SearchBar onChange={this.searchBarChange} value={this.state.searchText} />

        <SectionList
          keyExtractor={item => item.guid}
          initialNumToRender={20}
          maxToRenderPerBatch={40}
          renderItem={({ item }) => {
            return <UserItem key={item.guid} user={item}
              onPress={() => {
                this.controller.editUser(item)
              }} />
          }}
          renderSectionHeader={({ section }) => <SectionHeader section={section} />}
          sections={this.sections}
          style={{ flex: 1 }}
          refreshControl={<RefreshControl
            refreshing={stores.users.isRefreshing}
            onRefresh={stores.users.refresh}
          />}
        />

      </View>
    )
  }
}

@observer
class SectionHeader extends Component {
  render () {
    return (<View key={this.props.section.title} style={{
      backgroundColor: GlobalStyle.style.primaryColour,
      padding: 5
    }}>
      <Text style={{
        fontWeight: 'bold',
        fontSize: 16,
        color: GlobalStyle.style.primaryColourText
      }}>{this.props.section.title || 'No Title'}</Text>
    </View>)
  }
}

export class UserManagementController {
  @observable model = {
    users: [],
    newUser: {}
  }

  constructor (props) {
    this.props = props
  }

  startCreateUser = () => {
    this.model.newUser = stores.users.create()
    GlobalState.navigation.push({
      view: <UserDetailPage user={this.model.newUser} onSave={this.saveUser} />
    })
  }

  saveUser = (user) => {
    user.save()
  }

  deleteUser = (user) => {
    user.remove()
  }

  editUser (user) {
    GlobalState.navigation.push({
      view:
  <UserDetailPage user={user}
    canDelete
    onDelete={this.deleteUser}
    onSave={this.saveUser} />
    })
  }

  manageGroups = () => {
    GlobalState.navigation.resetTo({
      view: <GroupManagement />
    })
  }
}

export default UserManagement
export { UserItem, UserDetailPage, SectionHeader }
