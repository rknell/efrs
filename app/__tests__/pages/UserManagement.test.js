import React from 'react'
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'
import UserManagement, { SectionHeader, UserDetailPage, UserItem } from '../../app/pages/UserManagement'
import { shallow } from 'enzyme'
import fixtures from '../../__fixtures__'
import stores from '../../app/stores'
import { Alert } from 'react-native'
import GlobalState from '../../app/services/GlobalState'

import toJSON from 'enzyme-to-json'

jest.mock('mobx-react/native', () => require('mobx-react/custom'))

it('hasnt changed', () => {
  const tree = renderer.create(<UserManagement />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders a UserDetailPage (new user)', () => {
  const wrapper = shallow(<UserDetailPage user={fixtures.users.newUser} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('renders a UserDetailPage (setup user)', () => {
  const wrapper = shallow(<UserDetailPage user={fixtures.users.setupUser} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('renders a UserItem (new user)', () => {
  const wrapper = shallow(<UserItem user={fixtures.users.newUser} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('renders a UserItem (setup user)', () => {
  const wrapper = shallow(<UserItem user={fixtures.users.setupUser} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('renders a user without a name', () => {
  const wrapper = shallow(<UserItem user={{ ...fixtures.users.setupUser, firstName: '', lastName: '' }} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('renders a section header', () => {
  const wrapper = shallow(<SectionHeader section={{ title: 'Test' }} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('renders a section header with no title', () => {
  const wrapper = shallow(<SectionHeader section={{}} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('should get the users relationships', () => {
  const oldUserStore = stores.users.data

  stores.users.data = [
    {
      _id: 1,
      username: 'Test',
      relationships: [
        { name: 'supervisor', user: 2 }
      ]
    }, {
      _id: 2,
      lastName: 'User 2',
      username: 'User 2',
      relationships: []
    }
  ]

  const wrapper = shallow(<UserItem user={stores.users.data[0]} />)
  const instance = wrapper.instance()
  const output = instance.getRelationshipUser(2)
  expect(output).toBe('User 2')

  stores.users.data = oldUserStore
})

it('renders the last item', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  const wrapper = shallow(<UserItem user={user} isLast />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('handles save correctly, invalid user', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  const wrapper = shallow(<UserDetailPage user={user} />)
  const instance = wrapper.instance()
  const oldAlert = Alert.alert

  let alertShown = false
  Alert.alert = (title, message) => {
    expect(title).toBeDefined()
    expect(message).toBeDefined()
    alertShown = true
  }

  instance.onSaveButton()

  expect(alertShown).toBe(true)
  Alert.alert = oldAlert
})

it('handles save correctly, valid user', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  let onSaveCalled = false
  const wrapper = shallow(<UserDetailPage user={user} onSave={() => { onSaveCalled = true }} />)
  const instance = wrapper.instance()

  let popCalled = false
  GlobalState.navigation.pop = () => {
    popCalled = true
  }

  instance.state.isUserValid = true
  instance.onSaveButton()

  expect(popCalled).toBe(true)
  expect(onSaveCalled).toBe(true)
})

it('handles delete correctly', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  let onDeleteCalled = false
  const wrapper = shallow(<UserDetailPage user={user} canDelete onDelete={() => { onDeleteCalled = true }} />)
  const instance = wrapper.instance()

  let popCalled = false
  GlobalState.navigation.pop = () => {
    popCalled = true
  }

  instance.state.isUserValid = true
  instance.onDelete()

  expect(popCalled).toBe(true)
  expect(onDeleteCalled).toBe(true)
})

it('handles back, no changes', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  const wrapper = shallow(<UserDetailPage user={user} />)
  const instance = wrapper.instance()

  let popCalled = false
  GlobalState.navigation.pop = () => {
    popCalled = true
  }

  instance.handleBack()

  expect(popCalled).toBe(true)
})

it('handles back, user changes', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  const wrapper = shallow(<UserDetailPage user={user} />)
  const instance = wrapper.instance()
  user.lastName = 'User 4'
  wrapper.update()

  let popCalled = false
  GlobalState.navigation.pop = () => {
    popCalled = true
  }

  let alertShown = false
  Alert.alert = (title, message) => {
    expect(title).toBeDefined()
    expect(message).toBeDefined()
    alertShown = true
  }

  instance.handleBack()

  expect(popCalled).toBe(false)
  expect(alertShown).toBe(true)
})

it('handles onSaveBack', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  let onSaveCalled = false
  const wrapper = shallow(<UserDetailPage user={user} onSave={() => { onSaveCalled = true }} />)
  const instance = wrapper.instance()

  let popCalled = false
  GlobalState.navigation.pop = () => {
    popCalled = true
  }

  instance.onSaveBack()

  expect(popCalled).toBe(true)
  expect(onSaveCalled).toBe(true)
})

it('handles onNoSave', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  const wrapper = shallow(<UserDetailPage user={user} />)
  const instance = wrapper.instance()
  user.lastName = 'User 4'
  wrapper.update()

  let popCalled = false
  GlobalState.navigation.pop = () => {
    popCalled = true
  }

  instance.onNoSave()
  expect(popCalled).toBe(true)
  expect(user.lastName).toBe('User 2')
})

it('handles onValidationChanged', () => {
  const user = {
    _id: 2,
    lastName: 'User 2',
    username: 'User 2',
    relationships: []
  }
  const wrapper = shallow(<UserDetailPage user={user} />)
  const instance = wrapper.instance()
  expect(instance.state.isUserValid).toBe(false)
  instance.onValidationChanged(true)
  expect(instance.state.isUserValid).toBe(true)
})

it('filters correctly', () => {
  const wrapper = shallow(<UserManagement />)
  const instance = wrapper.instance()

  instance.state.searchText = 'TEST'
  expect(instance._meetsFilter('TEST')).toBe(true)
  expect(instance._meetsFilter('TESTS')).toBe(true)
  expect(instance._meetsFilter('STESTS')).toBe(true)
  expect(instance._meetsFilter('STE STS')).toBe(false)
  expect(instance._meetsFilter('TES')).toBe(false)
  expect(instance._meetsFilter('')).toBe(false)

  instance.state.searchText = ''
  expect(instance._meetsFilter('TEST')).toBe(true)
  expect(instance._meetsFilter('TESTS')).toBe(true)
  expect(instance._meetsFilter('STESTS')).toBe(true)
  expect(instance._meetsFilter('STE STS')).toBe(true)
  expect(instance._meetsFilter('TES')).toBe(true)
  expect(instance._meetsFilter('')).toBe(true)

  instance.state.searchText = 0
  expect(instance._meetsFilter('TEST')).toBe(true)
  expect(instance._meetsFilter('TESTS')).toBe(true)
  expect(instance._meetsFilter('STESTS')).toBe(true)
  expect(instance._meetsFilter('STE STS')).toBe(true)
  expect(instance._meetsFilter('TES')).toBe(true)
  expect(instance._meetsFilter('')).toBe(true)

  instance.state.searchText = '1'
  expect(instance._meetsFilter('TEST')).toBe(false)
  expect(instance._meetsFilter('TESTS')).toBe(false)
  expect(instance._meetsFilter('STESTS')).toBe(false)
  expect(instance._meetsFilter('STE STS')).toBe(false)
  expect(instance._meetsFilter('TES')).toBe(false)
  expect(instance._meetsFilter('')).toBe(false)
})

it('Creates sections correctly', () => {
  const wrapper = shallow(<UserManagement />)
  const instance = wrapper.instance()

  stores.users.data = [
    { lastName: 'Brian' },
    { lastName: 'WholeLotta' },
    { lastName: 'WholeLotta2' }
  ]

  expect(instance.sections.length).toBe(2)
})

it('Sorts sections correctly', () => {
  const wrapper = shallow(<UserManagement />)
  const instance = wrapper.instance()

  stores.users.data = [
    { lastName: 'WholeLotta' },
    { lastName: 'Brian' },
    { lastName: 'WholeLotta2' }
  ]

  expect(instance.sections.length).toBe(2)
  expect(instance.sections[0].title).toBe('B')
  expect(instance.sections[1].title).toBe('W')
})

it('Sorts users by their last name', () => {
  const wrapper = shallow(<UserManagement />)
  const instance = wrapper.instance()

  stores.users.data = [
    { lastName: 'WholeLotta' },
    { lastName: 'Whino' },
    { lastName: 'WholeLotta2' },
    { lastName: 'Wendel' }
  ]

  expect(instance.sections.length).toBe(1)
  expect(instance.sections[0].data[0]).toEqual({ lastName: 'Wendel' })
  expect(instance.sections[0].data[1]).toEqual({ lastName: 'Whino' })
})

it('handles a search bar change', () => {
  const wrapper = shallow(<UserManagement />)
  const instance = wrapper.instance()
  instance.searchBarChange('NEWVALUE')
  expect(instance.state.searchText).toBe('NEWVALUE')
})
