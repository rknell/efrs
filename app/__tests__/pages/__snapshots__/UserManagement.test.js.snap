// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hasnt changed 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "dodgerblue",
          "borderBottomColor": "rgb(106, 182, 255)",
          "borderBottomWidth": 0,
          "elevation": 1,
        }
      }
    >
      <View
        style={
          Object {
            "height": 22,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "padding": 10,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "padding": 6,
                "paddingLeft": 16,
                "paddingRight": 16,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "rgb(255, 255, 255)",
                    "fontSize": 18,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <Text
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "rgb(255, 255, 255)",
              "flex": 1,
              "fontSize": 20,
              "fontWeight": "300",
            }
          }
        >
          User Management
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f3f3f3",
          "borderBottomWidth": 0,
          "borderColor": "rgb(205, 205, 205)",
          "borderTopWidth": 0,
          "flexDirection": "row",
          "paddingLeft": 16,
          "paddingRight": 16,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginRight": 10,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "borderColor": "rgba(255,255,255,0.3)",
                "borderRightWidth": 1,
                "marginRight": 10,
                "padding": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "rgb(39, 39, 39)",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
               
              Add user
               
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "borderColor": "rgba(255,255,255,0.3)",
                "borderRightWidth": 0,
                "marginRight": 0,
                "padding": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "rgb(39, 39, 39)",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
               
              Manage Groups
               
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={20}
    keyExtractor={[Function]}
    maxToRenderPerBatch={40}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    renderItem={[Function]}
    renderSectionHeader={[Function]}
    scrollEventThrottle={50}
    sections={Array []}
    stickyHeaderIndices={Array []}
    stickySectionHeadersEnabled={true}
    style={
      Object {
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;

exports[`renders a UserDetailPage (new user) 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <NavBar
    handlePop={[Function]}
    isMenuDisabled={true}
    pageActions={
      Array [
        Object {
          "icon": "save",
          "label": "Save Changes",
          "onPress": [Function],
        },
      ]
    }
    title="User Details"
  />
  <UserDetailView
    onValidationChanged={[Function]}
    user={
      Object {
        "__v": 0,
        "_id": "5a36fa6aac023a4b3bae2f22",
        "company": "5a36fa6aac023a4b3bae2f20",
        "createdAt": "2017-12-17T23:14:50.373Z",
        "domain": "example.com.au",
        "email": "user12@example.com",
        "firstName": "Admin",
        "isSetup": false,
        "lastName": "User",
        "password": "$2a$08$gtSMlrmdhH0zOeeYJzPh.egq2I65fD9WFV5PypJfNWqikrZOzFQJq",
        "pushTokens": Array [],
        "relationships": Array [],
        "updatedAt": "2017-12-17T23:14:50.390Z",
        "username": "user12@example.com",
      }
    }
  />
</View>
`;

exports[`renders a UserDetailPage (setup user) 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <NavBar
    handlePop={[Function]}
    isMenuDisabled={true}
    pageActions={
      Array [
        Object {
          "icon": "save",
          "label": "Save Changes",
          "onPress": [Function],
        },
      ]
    }
    title="User Details"
  />
  <UserDetailView
    onValidationChanged={[Function]}
    user={
      Object {
        "__v": 0,
        "_id": "5a36fa6aac023a4b3bae2f22",
        "company": "5a36fa6aac023a4b3bae2f20",
        "createdAt": "2017-12-17T23:14:50.373Z",
        "domain": "example.com.au",
        "email": "user12@example.com",
        "firstName": "Admin",
        "isSetup": true,
        "lastName": "User",
        "password": "$2a$08$gtSMlrmdhH0zOeeYJzPh.egq2I65fD9WFV5PypJfNWqikrZOzFQJq",
        "pushTokens": Array [],
        "relationships": Array [],
        "updatedAt": "2017-12-17T23:14:50.390Z",
        "username": "user12@example.com",
      }
    }
  />
</View>
`;

exports[`renders a UserItem (new user) 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
>
  <View
    style={
      Object {
        "borderBottomColor": "#e7eaec",
        "borderBottomWidth": 1,
        "padding": 5,
        "paddingLeft": 8,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        <Text>
          User
          , 
        </Text>
        <Text>
          Admin
        </Text>
      </Text>
    </View>
    <Text>
      user12@example.com
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`renders a UserItem (setup user) 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
>
  <View
    style={
      Object {
        "borderBottomColor": "#e7eaec",
        "borderBottomWidth": 1,
        "padding": 5,
        "paddingLeft": 8,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        <Text>
          User
          , 
        </Text>
        <Text>
          Admin
        </Text>
      </Text>
    </View>
    <Text>
      user12@example.com
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`renders a section header 1`] = `
<View
  key="Test"
  style={
    Object {
      "backgroundColor": "dodgerblue",
      "padding": 5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(255, 255, 255)",
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
  >
    Test
  </Text>
</View>
`;

exports[`renders a section header with no title 1`] = `
<View
  style={
    Object {
      "backgroundColor": "dodgerblue",
      "padding": 5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(255, 255, 255)",
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
  >
    No Title
  </Text>
</View>
`;

exports[`renders a user without a name 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
>
  <View
    style={
      Object {
        "borderBottomColor": "#e7eaec",
        "borderBottomWidth": 1,
        "padding": 5,
        "paddingLeft": 8,
      }
    }
  >
    <Text>
      user12@example.com
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`renders the last item 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
>
  <View
    style={
      Object {
        "borderBottomColor": "transparent",
        "borderBottomWidth": 1,
        "padding": 5,
        "paddingLeft": 8,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        <Text>
          User 2
          , 
        </Text>
      </Text>
    </View>
    <Text>
      User 2
    </Text>
  </View>
</TouchableOpacity>
`;
